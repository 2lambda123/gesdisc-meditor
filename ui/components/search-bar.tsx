import Router from 'next/router'
import { useQuery } from '@apollo/react-hooks'
import { useState, useEffect } from 'react'
import gql from 'graphql-tag'
import Button from 'react-bootstrap/Button'
import Form from 'react-bootstrap/Form'
import Select from 'react-select'
import styles from './search-bar.module.css'
import { withApollo } from '../lib/apollo'
import ModelIcon from './model-icon'
import { MdSearch } from 'react-icons/md'
import { useInput } from '../lib/use-input.hook'
import { useDebounce } from '../lib/use-debounce.hook'

/**
 * queries all models for display in the select dropdown
 */
const ALL_MODELS_QUERY = gql`
    {
        models {
            name
            icon {
                name
                color
            }
        }
    }
`

/**
 * returns an option to be rendered in the Model list
 * @param model
 */
const formatOptionLabel = (model) => (
    <div style={{ display: 'flex', alignItems: 'center' }}>
        <ModelIcon name={model.icon.name} color={model.icon.color} />
        <div style={{ marginLeft: 5 }}>{ model?.name }</div>
    </div>
)

const SearchBar = ({ modelName, onInput }) => {
    const { data } = useQuery(ALL_MODELS_QUERY)

    const { value, bind, reset } = useInput('')
    const [selectedModel, setSelectedModel] = useState(null)
    const [options, setOptions] = useState([])
    const searchTerm = useDebounce(value, 200)
 
    /**
     * when models are returned from the API, set them in the select and then select the matching model
     */
    useEffect(() => {
        setOptions(data?.models || [])
        setSelectedModel(data?.models.find(model => model.name === modelName))
    }, [data])

    /**
     * notify when search changes
     */
    useEffect(() => {
        onInput(searchTerm)
    }, [searchTerm])

    /**
     * route to the requested model when a different one is selected
     * @param selectedModel 
     */
    function handleModelChange(selectedModel) {
        setSelectedModel(selectedModel)
        reset()
        Router.push('/[modelName]', `/${selectedModel.name}`)
    }

    return (
        <div className={styles.container}>
            <Form className={styles.form}>
                <Select
                    instanceId="search-bar-select"
                    value={selectedModel}
                    formatOptionLabel={formatOptionLabel}
                    onChange={handleModelChange}
                    options={options}
                    className="model-select"
                    classNamePrefix="model-select"
                    isSearchable={false}
                    placeholder=""
                />

                <input
                    type="text"
                    name="search"
                    className={`form-control ${styles.input}`}
                    placeholder={`Search ${modelName} for...`}
                    {...bind}
                />

                <Button className={styles.button}>
                    <MdSearch />
                </Button>
            </Form>
        </div>
    )
}

export default withApollo({ ssr: true })(SearchBar)
