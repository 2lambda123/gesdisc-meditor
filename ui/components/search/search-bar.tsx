import Router from 'next/router'
import { useQuery } from '@apollo/react-hooks'
import { useState, useEffect } from 'react'
import gql from 'graphql-tag'
import Button from 'react-bootstrap/Button'
import Form from 'react-bootstrap/Form'
import Select from 'react-select'
import styles from './search-bar.module.css'
import { withApollo } from '../../lib/apollo'
import ModelIcon from '../model-icon'
import { MdSearch } from 'react-icons/md'
import { useInput } from '../../lib/use-input.hook'
import { useDebounce } from '../../lib/use-debounce.hook'

/**
 * queries all models for display in the select dropdown
 */
const ALL_MODELS_QUERY = gql`
    {
        models {
            name
            icon {
                name
                color
            }
        }
    }
`

/**
 * returns an option to be rendered in the Model list
 * @param model
 */
const formatOptionLabel = (model) => (
    <div style={{ display: 'flex', alignItems: 'center' }}>
        <ModelIcon name={model.icon.name} color={model.icon.color} />
        <div style={{ marginLeft: 5 }}>{ model?.name }</div>
    </div>
)

const SearchBar = ({ model, modelName, initialInput = '', onInput }) => {
    const { data } = useQuery(ALL_MODELS_QUERY)

    const { value, bind, reset } = useInput(initialInput)
    const [selectedModel, setSelectedModel] = useState(null)
    const [options, setOptions] = useState([])
    const searchTerm = useDebounce(value, 200)

    /**
     * set models as selectable options in the dropdown and then select the matching model
     * @param models 
     */
    function updateSelectedModelsDropdown(models) {
        if (!models) return
        setOptions(models)
        setSelectedModel(models.find(model => model.name === modelName))
    }

    /**
     * default the dropdown to just the selected model
     */
    useEffect(() => {
        // don't update dropdown if we've already set it up
        if (options.length || !model) return

        updateSelectedModelsDropdown([model])
    }, [model, options])
 
    /**
     * when models are returned from the API, set them in the dropdown
     */
    useEffect(() => {
        updateSelectedModelsDropdown(data?.models || [])
    }, [data])

    /**
     * notify when search changes
     */
    useEffect(() => {
        onInput(searchTerm)
    }, [searchTerm])

    /**
     * route to the requested model when a different one is selected
     * @param selectedModel 
     */
    function handleModelChange(selectedModel) {
        setSelectedModel(selectedModel)
        reset()
        Router.push('/[modelName]', `/${selectedModel.name}`)
    }

    return (
        <div className={styles.container}>
            <Form className={styles.form}>
                <Select
                    instanceId="search-bar-select"
                    value={selectedModel}
                    formatOptionLabel={formatOptionLabel}
                    onChange={handleModelChange}
                    options={options}
                    className="model-select"
                    classNamePrefix="model-select"
                    isSearchable={false}
                    placeholder=""
                />

                <input
                    type="text"
                    name="search"
                    className={`form-control ${styles.input}`}
                    placeholder={`Search ${modelName} for...`}
                    {...bind}
                />

                <Button className={styles.button}>
                    <MdSearch />
                </Button>
            </Form>
        </div>
    )
}

export default withApollo({ ssr: true })(SearchBar)
