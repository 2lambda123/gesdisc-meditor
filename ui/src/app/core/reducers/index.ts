import { createSelector, createFeatureSelector, ActionReducerMap } from '@ngrx/store';
import * as fromContentTypes from './content-types';

export interface ContentTypesState {
  contentTypes: fromContentTypes.State;
}

export interface State {
  contentTypes: ContentTypesState;
}

export const reducers = {
  contentTypes: fromContentTypes.reducer
};

export const selectContentTypesState = createFeatureSelector<ContentTypesState>('contentTypes');


export const getContentTypesEntitiesState = createSelector(
  selectContentTypesState,
  state => state.contentTypes
);

export const {
  selectIds: getContentTypeIds,
  selectEntities: getContentTypeEntities,
  selectAll: getAllContentTypes,
  selectTotal: getTotalContentTypes,
} = fromContentTypes.adapter.getSelectors(getContentTypesEntitiesState);
// export const selectContentTypesIds = createSelector(selectContentTypesState, fromContentTypes.selectContentTypesIds);
// export const selectContentTypesEntities = createSelector(selectContentTypesState, fromContentTypes.selectContentTypesEntities);
// export const selectAllContentTypes = createSelector(selectContentTypesState, fromContentTypes.selectAllContentTypes);
// export const selectContentTypesTotal = createSelector(selectContentTypesState, fromContentTypes.selectContentTypesTotal);
// export const selectCurrentUserId = createSelector(selectUserState, fromContentTypes.getSelectedUserId);

// export const selectCurrentUser = createSelector(
//   selectUserEntities,
//   selectCurrentUserId,
//   (userEntities, userId) => userEntities[userId]
// );
