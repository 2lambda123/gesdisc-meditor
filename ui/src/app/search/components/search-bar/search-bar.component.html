<div fxLayout="row wrap" fxLayoutAlign="center center" class="search-bar-container">
  <mat-form-field>
	  <mat-select [(ngModel)]="selected">
	  	<mat-select-trigger *ngIf="selected">
		  	<i class="icon-badge icon-badge-xs fa {{selected.icon.name}} align-middle" [style.background-color]="selected.icon.color"></i>
		    {{selected.name}}
      </mat-select-trigger>
	    <mat-option *ngFor="let contentType of contentTypes$ | async" [value]="contentType">
	    	<i class="icon-badge icon-badge-xs fa {{contentType.icon.name}} align-middle" [style.background-color]="contentType.icon.color"></i>
	    	{{contentType.name}}
	    </mat-option>
	  </mat-select>
	</mat-form-field>
	<mat-form-field class="search-bar-input">
		<input matInput placeholder="Search..." type="text" name="search" >
	</mat-form-field>
</div>
