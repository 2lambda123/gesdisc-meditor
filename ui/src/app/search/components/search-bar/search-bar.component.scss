.search-bar-input {
	width: 100%;
	font-size: 18px;
}

.search-bar-select {
	width: 100%;
	font-size: 18px;
}


