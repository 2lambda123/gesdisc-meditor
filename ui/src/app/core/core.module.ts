import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { FlexLayoutModule } from '@angular/flex-layout';
// CONTAINERS
import { MainComponent } from './containers/main';
import { NotFoundPageComponent } from './containers/not-found-page';
import { SplashPageContainer } from './containers/splash-page/splash-page.container';
// COMPONENTS
import { ContentTypeBoxComponent } from './components/content-type-button/content-type-button.component';
import { ContentTypeSplashComponent } from './components/splash-box/splash-box.component';
import { LayoutComponent } from './components/layout';
import { NavItemComponent } from './components/nav-item';
import { SidenavComponent } from './components/sidenav';
import { ToolbarComponent } from './components/toolbar';
import { MaterialModule } from '../material';

import { AngularFontAwesomeModule } from 'angular-font-awesome';

import { AuthModule } from '../auth/auth.module';

export const COMPONENTS = [
  MainComponent,
  NotFoundPageComponent,
  SplashPageContainer,
  ContentTypeBoxComponent,
  ContentTypeSplashComponent,
  LayoutComponent,
  NavItemComponent,
  SidenavComponent,
  ToolbarComponent,
];

@NgModule({
  imports: [CommonModule, RouterModule, MaterialModule, FlexLayoutModule, AngularFontAwesomeModule, AuthModule.forRoot()],
  declarations: COMPONENTS,
  exports: COMPONENTS
})

export class CoreModule {
  static forRoot() {
    return {
      ngModule: CoreModule
    };
  }
}
