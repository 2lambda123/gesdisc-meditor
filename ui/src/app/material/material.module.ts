import { NgModule } from '@angular/core';

import {
  MatInputModule,
  MatCardModule,
  MatButtonModule,
  MatSidenavModule,
  MatListModule,
  MatIconModule,
  MatToolbarModule,
  MatProgressSpinnerModule,
  MatDatepickerModule,
  MatCheckboxModule,
  MatSelectModule
} from '@angular/material';

@NgModule({
  imports: [
    MatInputModule,
    MatCardModule,
    MatButtonModule,
    MatSidenavModule,
    MatListModule,
    MatIconModule,
    MatToolbarModule,
    MatProgressSpinnerModule,
    MatDatepickerModule,
    MatCheckboxModule,
  	MatSelectModule
  ],
  exports: [
    MatInputModule,
    MatCardModule,
    MatButtonModule,
    MatSidenavModule,
    MatListModule,
    MatIconModule,
    MatToolbarModule,
    MatProgressSpinnerModule,
    MatDatepickerModule,
    MatCheckboxModule,
  	MatSelectModule
  ],
})
export class MaterialModule {}
