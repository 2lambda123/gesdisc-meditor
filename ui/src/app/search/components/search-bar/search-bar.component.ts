import { Component, Input, Output, EventEmitter, OnInit, OnChanges } from '@angular/core'
import { FormControl } from '@angular/forms'
import { ModelCatalogEntry } from '../../../service'

@Component({
    selector: 'med-search-bar',
    templateUrl: './search-bar.component.html',
    styleUrls: ['./search-bar.component.scss'],
})
export class SearchBarComponent implements OnInit {
    @Input() models: ModelCatalogEntry[]
    @Input() selectedModel: ModelCatalogEntry
    @Input() searchTerm: string
    @Output() selectionChanged = new EventEmitter<string>()
    @Output() searchChanged = new EventEmitter<string>()

    query: string
    modelcontrol = new FormControl()

    ngOnInit() {
        this.query = this.searchTerm
        this.setModelControl()
    }

    ngOnChanges() {
        this.query = this.searchTerm
        this.setModelControl()
    }

    onSearchChange() {
        this.searchChanged.emit(this.query)
    }

    setModelControl() {
        const i = this.models.map(m => m.name).indexOf(this.selectedModel.name)
        this.modelcontrol.setValue(this.models[i])
    }
}
