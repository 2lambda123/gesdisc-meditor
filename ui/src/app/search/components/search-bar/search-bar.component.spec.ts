import { ComponentFixture, TestBed  } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { MatSelectModule, MatIconModule, MatInputModule } from '@angular/material';
import { SearchBarComponent } from './search-bar.component';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';

describe('SearchBarComponent', () => {
	let fixture: ComponentFixture<SearchBarComponent>;
	let instance: SearchBarComponent;

	beforeEach(() => {
		TestBed.configureTestingModule({
			imports: [
				FormsModule,
				ReactiveFormsModule,
				NoopAnimationsModule,
				MatSelectModule,
				MatInputModule,
				MatIconModule
			],
			declarations: [
				SearchBarComponent
			]
		});

		fixture = TestBed.createComponent(SearchBarComponent);
		instance = fixture.componentInstance;

	});

	describe('with no data', () => {
		it('should not have any initial input', () => {
			expect(instance.models).toBeUndefined();
			expect(instance.selectedModel).toBeUndefined();
		});
	});

	describe('with mock data', () => {

		it('should compile with minimum input', () => {
			instance.models = [
				{
					name: 'test',
					description: 'test'
				},
				{
					name: 'test1',
					description: 'test1'
				}
			]
			instance.selectedModel = {
				name: 'test',
				description: 'test'
			}
			fixture.detectChanges();
			expect(fixture).toMatchSnapshot();
		});
	});
});
