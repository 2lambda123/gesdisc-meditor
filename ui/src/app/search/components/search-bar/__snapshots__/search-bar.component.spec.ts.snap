// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBarComponent with mock data should compile with minimum input 1`] = `
<med-search-bar
  modelcontrol={[Function FormControl]}
  models={[Function Array]}
  searchChanged={[Function EventEmitter]}
  selectedModel={[Function Object]}
  selectionChanged={[Function EventEmitter]}
>
  <div
    class="search-bar-container"
  >
    <mat-select
      aria-disabled="false"
      aria-invalid="false"
      aria-multiselectable="false"
      aria-required="false"
      class="search-bar-select mat-select ng-tns-c3-0 ng-untouched ng-pristine ng-valid"
      id="mat-select-1"
      ng-reflect-form="[object Object]"
      role="listbox"
      tabindex="0"
    >
      <div
        aria-hidden="true"
        cdk-overlay-origin=""
        class="mat-select-trigger"
      >
        <div
          class="mat-select-value"
          ng-reflect-ng-switch="true"
        >
          
          <span
            class="mat-select-placeholder ng-tns-c3-0 ng-star-inserted"
          >
            Â 
          </span>
          
        </div>
        <div
          class="mat-select-arrow-wrapper"
        >
          <div
            class="mat-select-arrow"
          />
        </div>
      </div>
      
    </mat-select>
    <input
      aria-invalid="false"
      aria-required="false"
      class="search-bar-input mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-valid"
      id="mat-input-1"
      matinput=""
      name="search"
      ng-reflect-name="search"
      ng-reflect-placeholder="Search test for..."
      ng-reflect-type="text"
      placeholder="Search test for..."
      type="text"
    />
    <button
      class="search-bar-submit"
      color="primary"
      mat-flat-button=""
    >
      <mat-icon
        aria-hidden="true"
        class="mat-icon material-icons"
        matsuffix=""
        role="img"
      >
        search
      </mat-icon>
    </button>
  </div>
</med-search-bar>
`;
